source/opt/passes.h

```diff
@@ -74,6 +79,132 @@ class EliminateDeadConstantPass : public Pass {
   bool Process(ir::Module*) override;
 };
 
+// The optimization pass for folding SpecConstantOp and SpecConstantComposite
+// instructions. After freezing, spec constants defined with Spec ID are
+// specialized into front-end constants. This pass folds the SpecConstantOp
+// instructions and SpecConstantComposite instructions that depend on those
+// frozen spec constants.
+class FoldSpecConstantOpAndCompositePass : public Pass {
+ public:
+   // Data structure to record the in-operand of front-end constants.
+   // ScalarValue stores the value of of scalar constants. CompositeValue
+   // stores a vector of its component ids. Value is a abstract class of
+   // ScalarValue and CompositeValue.
+  struct ScalarValue;
+  struct CompositeValue;
+  struct Value {
+    virtual ScalarValue* AsScalar() { return nullptr; }
+    virtual CompositeValue* AsComposite() { return nullptr; }
+  };
+  struct ScalarValue : public Value {
+    ScalarValue(uint32_t w) : word(w) {}
+    ScalarValue(const ScalarValue* other) : word(other->word) {}
+    ScalarValue* AsScalar() override { return this; }
+    uint32_t word;
+  };
+  struct CompositeValue : public Value {
+    CompositeValue() : nested() {}
+    CompositeValue(const CompositeValue* other) : nested(other->nested) {}
+    CompositeValue* AsComposite() override { return this; }
+    std::vector<uint32_t> nested;  // Ids of the nested constant.
+  };
+
+  FoldSpecConstantOpAndCompositePass()
+      : Pass(),
+        max_id_(0),
+        def_use_mgr_(nullptr),
+        type_mgr_(nullptr),
+        id_to_const_val_() {}
+  const char* name() const override { return "fold-spec-const-op-composite"; }
+  bool Process(ir::Module* module) override {
+    Initialize(module);
+    return ProcessImpl(module);
+  };
+
+ private:
+  // Initializes the type manager, def-use manager and get the maximum id used
+  // in the module.
+  void Initialize(ir::Module* module) {
+    type_mgr_.reset(new analysis::TypeManager(*module));
+    def_use_mgr_.reset(new analysis::DefUseManager(module));
+    module->ForEachInst([&](ir::Instruction* inst) {
```

Comment:
Hmm, I think we can get this from `type_mgr_.id_to_defs()`. `ForEachInst` is not a lightweight call.